core/services/ocr2/plugins/ocr2keeper/evm21/registry_check_pipeline.go

```diff
@@ -40,8 +41,16 @@ func (r *EvmRegistry) CheckUpkeeps(ctx context.Context, keys ...ocr2keepers.Upke
 		}
 	}
 
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer wg.Wait()
+
 	chResult := make(chan checkResult, 1)
-	go r.doCheck(ctx, keys, chResult)
+
+	go func() {
+		defer wg.Done()
+		r.doCheck(ctx, keys, chResult)
+	}()
 
 	select {
 	case rs := <-chResult:
```

Comment:
A quick mitigation could be to return error if `ctx.Err() != nil` that would protect us in cases the context was done.

Reply:
Do we need to do that in addition to using the thread control?

Reply:
Within `doCheck` we won't stop in case the context (threadCtrl ctx) was done, you should fix that within the internal functions calls, e.g. `streamsLookup()` spawns goroutines for `doLookup` w/o threadCtrl.

Let's avoid `ctx.Err() != nil` and fix `streamsLookup` to use threadCtrl.

Reply:
But isn't doLookup already sync'd with a waitGroup?

Reply:
ThreadCtrl is also about stopping any running goroutine (and wait for them to stop), which we don't have right now